File: 003/code.go
//The prime factors of 13195 are 5, 7, 13 and 29.

//What is the largest prime factor of the number 600851475143 ?

package main

import "fmt"

func main() {
	factors := primeFactors(600851475143)
	fmt.Println(factors)
	fmt.Println(max(factors))
}

func firstPrimeFactor(number int) int {
	n := 2
	for n < number {
		if number%n == 0 {
			return n
		}
		n++
	}
	return number
}

func primeFactors(number int) []int {
	var result []int
	for {
		factor := firstPrimeFactor(number)
		result = append(result, factor)
		if number == factor {
			return result
		}
		number = number / factor
	}
}

func max(numbers []int) int {
	result := numbers[0]
	for _, value := range numbers {
		if value > result {
			result = value
		}
	}
	return result
}

File: 003/code.jl
# The prime factors of 13195 are 5, 7, 13 and 29.

# What is the largest prime factor of the number 600851475143 ?

function firstPrimeFactor(n)
  num = 2
  while num <= n
    if mod(n, num) == 0
      return num
    end
    num += 1
  end
end

function primeFactors(n)
  result = []
  while true
    p = firstPrimeFactor(n)
    push!(result, p)
    if p == n
      return result
    end
    n = div(n, p)
  end
end

function maxValue(arr)
  result = arr[1]
  for num in arr
    if num > result
      result = num
    end
  end
  result
end

println(maxValue(primeFactors(600851475143)))
